File: quiz_app_flutter/lib/domain/usecase/result/getresultdata_usecase.dart
import 'dart:convert';

import 'package:quiz_app_flutter/data/cache/cache_manager.dart';
import 'package:quiz_app_flutter/domain/entities/result/result_data.dart';

class GetResultDataUseCase {
  final CacheManager cacheManager;

  GetResultDataUseCase(this.cacheManager);

  Stream<ResultData?> call(String key) async* {
    final jsonString = cacheManager.getResult(key);
    if (jsonString != null) {
      final Map<String, dynamic> jsonMap = json.decode(jsonString);
      yield ResultData.fromJson(jsonMap);
    } else {
      yield null;
    }
  }
}

File: quiz_app_kmp/composeApp/src/commonMain/kotlin/com/tnm/quizappkmp/domain/usecase/result/GetResultDataUseCase.kt
package com.tnm.quizappkmp.domain.usecase.result

import com.tnm.quizappkmp.data.cache.CacheManager
import com.tnm.quizappkmp.domain.model.result.ResultData
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.serialization.json.Json

class GetResultDataUseCase(
    private val cacheManager: CacheManager,
    private val json: Json
) {
    operator fun invoke(key: String): Flow<ResultData?> = flow {
        val result = cacheManager.getResult(key)?.let {
            json.decodeFromString<ResultData>(it)
        }
        emit(result)
    }
}